src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -300,6 +458,16 @@ protected virtual void OnBeforeNonRoslynDocumentWindowShow(IVsWindowFrame frame,
 
         private IList<DocumentId> GetDocumentIdsFromDocCookie(uint docCookie)
         {
+            lock (_gate)
+            {
+                return GetDocumentIdsFromDocCookie_NoLock(docCookie);
+            }
+        }
+
+        private IList<DocumentId> GetDocumentIdsFromDocCookie_NoLock(uint docCookie)
+        {
+            AssertIsForeground();
```

Comment:
Seems better pattern would be to move this assert outside the lock at start of GetDocumentIdsFromDocCookie